### ISSUE TITLE
Phase 45.2: PruningEngine — Network Pruning & Lottery Ticket Identification

### ISSUE THREAD

**Issue by web3guru888** [web3guru888/asi-build#890]
### Overview
Implements comprehensive neural network pruning framework supporting structured and unstructured pruning strategies, lottery ticket identification, movement pruning, and gradual magnitude pruning. Enables dramatic model size reduction while maintaining task performance.

#### 1. Magnitude-Based Pruning (Han et al. 2015)
- Element-wise (unstructured) weight pruning by absolute magnitude
- Global vs layer-wise pruning thresholds
- One-shot and iterative pruning schedules
- Weight rewinding after pruning (lottery ticket approach)

#### 2. Structured Pruning
- Filter pruning: remove entire convolutional filters
- Channel pruning: remove input/output channels
- Layer pruning: remove entire layers from deep networks
- Geometric median-based filter selection (He et al. 2019)
- Taylor expansion importance estimation

#### 3. Lottery Ticket Hypothesis (Frankle & Carlin 2019)
- Iterative magnitude pruning (IMP) with weight rewinding
- Late rewinding to early training epoch (Frankle et al. 2020)
- Mask identification and subnetwork extraction
- Lottery ticket transfer across tasks/datasets
- Supermask discovery (Zhou et al. 2019)

#### 4. Movement Pruning (Sanh et al. 2020)
- First-order gradient-based importance scoring
- Pruning weights that move toward zero during fine-tuning
- Soft movement pruning with straight-through estimator
- Task-specific pruning during transfer learning

#### 5. Gradual Magnitude Pruning (Zhu & Gupta 2017)
- Cubic sparsity schedule: `s_t = s_f + (s_i - s_f)(1 - (t-t_0)/(t_f-t_0))³`
- Automated schedule generation for target sparsity
- Layer-wise sparsity allocation based on sensitivity
- Pruning with fine-tuning interleaving

### API Design

```python
class PruningEngine:
    def __init__(self, config: PruningConfig):
        """Initialize pruning engine with strategy configuration."""
    
    def prune(
        self,
        model: nn.Module,
        strategy: str = "magnitude",  # magnitude|structured|lottery|movement|gradual
        sparsity_target: float = 0.9,
        schedule: str = "one_shot"  # one_shot|iterative|cubic
    ) -> PrunedModel:
        """Apply pruning strategy to model."""
    
    def find_lottery_ticket(
        self,
        model: nn.Module,
        train_loader: DataLoader,
        target_sparsity: float = 0.9,
        rewinding_epoch: int = 0,
        rounds: int = 15
    ) -> Tuple[nn.Module, PruningMask]:
        """Identify winning lottery ticket via iterative magnitude pruning."""
    
    def sensitivity_analysis(
        self,
        model: nn.Module,
        val_loader: DataLoader,
        sparsity_levels: List[float] = [0.1, 0.3, 0.5, 0.7, 0.9]
    ) -> SensitivityReport:
        """Analyze per-layer pruning sensitivity."""
    
    def movement_prune(
        self,
        model: nn.Module,
        train_loader: DataLoader,
        sparsity_target: float = 0.9
    ) -> PrunedModel:
        """Apply movement pruning during fine-tuning."""
```

### Acceptance Criteria
- [ ] Unstructured magnitude pruning achieving 90%+ sparsity
- [ ] Structured filter/channel pruning with actual speedup measurement
- [ ] Lottery ticket identification with weight rewinding
- [ ] Movement pruning for fine-tuning scenarios
- [ ] Cubic gradual pruning schedule implementation
- [ ] Per-layer sensitivity analysis with visualization
- [ ] Accuracy within 1% of dense model at 80% sparsity on benchmarks
- [ ] Unit tests with >90% coverage

### References
- Han et al. (2015) "Learning both Weights and Connections for Efficient Neural Networks"
- Frankle & Carlin (2019) "The Lottery Ticket Hypothesis"
- Sanh et al. (2020) "Movement Pruning: Adaptive Sparsity during Fine-Tuning"
- Zhu & Gupta (2017) "To Prune, or Not to Prune"
- He et al. (2019) "Filter Pruning via Geometric Median"

> Part of [Phase 45 Planning](#888)
